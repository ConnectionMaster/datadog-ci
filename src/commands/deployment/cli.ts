import {DeploymentMarkCommand} from './mark'

module.exports = [DeploymentMarkCommand]