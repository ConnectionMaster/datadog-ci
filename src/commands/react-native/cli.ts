import {UploadCommand} from './upload'
import {XCodeCommand} from './xcode'

module.exports = [UploadCommand, XCodeCommand]
