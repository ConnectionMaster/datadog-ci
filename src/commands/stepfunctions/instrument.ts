import {CloudWatchLogsClient} from '@aws-sdk/client-cloudwatch-logs'
import {IAMClient} from '@aws-sdk/client-iam'
import {SFNClient} from '@aws-sdk/client-sfn'
import {Command} from 'clipanion'

import {
  createLogGroup,
  enableStepFunctionLogs,
  describeStateMachine,
  listTagsForResource,
  putSubscriptionFilter,
  tagResource,
  attachPolicyToStateMachineIamRole,
  createLogsAccessPolicy,
} from './awsCommands'
import {TAG_VERSION_NAME} from './constants'
import {
  buildLogGroupName,
  buildArn,
  buildSubscriptionFilterName,
  isValidArn,
  parseArn,
  getStepFunctionLogGroupArn,
} from './helpers'

const cliVersion = require('../../../package.json').version

export class InstrumentStepFunctionsCommand extends Command {
  public static usage = Command.Usage({
    description: 'Subscribe Step Function Log Groups to a Datadog Forwarder',
    details: '--step-function expects a Step Function ARN\n--forwarder expects a Lambda ARN',
    examples: [
      [
        'View and apply changes to subscribe a Step Function Log Group to a Datadog Forwarder',
        'datadog-ci stepfunctions instrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder --env dev --service example-service',
      ],
      [
        'View changes to subscribe a Step Function Log Group to a Datadog Forwarder',
        'datadog-ci stepfunctions instrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder --env dev --service example-service --dry-run',
      ],
      [
        'View and apply changes to subscribe multiple Step Function Log Groups to a Datadog Forwarder',
        'datadog-ci stepfunctions instrument --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction1 --step-function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction2 --forwarder arn:aws:lambda:us-east-1:000000000000:function:ExampleDatadogForwarder --env dev --service example-service',
      ],
    ],
  })

  private dryRun = false
  private environment?: string
  private forwarderArn!: string
  private service?: string
  private stepFunctionArns: string[] = []

  public async execute() {
    let validationError = false
    if (typeof this.forwarderArn !== 'string') {
      this.context.stdout.write('[Error] `--forwarder` is required\n')
      validationError = true
    } else if (!isValidArn(this.forwarderArn)) {
      this.context.stdout.write(`[Error] Invalid arn format for \`--forwarder\` ${this.forwarderArn}\n`)
      validationError = true
    }

    // remove duplicate step function arns
    const stepFunctionArns = [...new Set(this.stepFunctionArns)]

    if (stepFunctionArns.length === 0) {
      this.context.stdout.write(`[Error] must specify at least one --step-function\n`)
      validationError = true
    }

    for (const stepFunctionArn of stepFunctionArns) {
      if (!isValidArn(stepFunctionArn)) {
        this.context.stdout.write(`[Error] invalid arn format for --step-function ${stepFunctionArn}\n`)
        validationError = true
      }
    }

    if (validationError) {
      return 1
    }

    let hasChanges = false

    // loop over step functions passed as parameters and generate a list of requests to make to AWS for each step function
    for (const stepFunctionArn of stepFunctionArns) {
      // use region from the step function arn to make requests to AWS
      const arnObject = parseArn(stepFunctionArn)
      const region = arnObject.region
      const cloudWatchLogsClient = new CloudWatchLogsClient({region})
      const stepFunctionsClient = new SFNClient({region})
      const iamClient = new IAMClient({region})

      let describeStateMachineCommandOutput
      try {
        describeStateMachineCommandOutput = await describeStateMachine(stepFunctionsClient, stepFunctionArn)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(`\n[Error] ${err.message}. Unable to describe state machine ${stepFunctionArn}\n`)
        } else {
          this.context.stdout.write(`\n[Error] ${err}. Unable to describe state machine ${stepFunctionArn}\n`)
        }

        return 1
      }

      let listStepFunctionTagsResponse
      try {
        listStepFunctionTagsResponse = await listTagsForResource(stepFunctionsClient, stepFunctionArn)
      } catch (err) {
        if (err instanceof Error) {
          this.context.stdout.write(
            `\n[Error] ${err.message}. Unable to fetch tags for Step Function ${stepFunctionArn}\n`
          )
        } else {
          this.context.stdout.write(`\n[Error] ${err}. Unable to fetch tags for Step Function ${stepFunctionArn}\n`)
        }

        return 1
      }

      const stepFunctionTagsToAdd = []

      // if env and service tags are not already set on step function, set these tags using the values passed as parameters
      const hasEnvTag = listStepFunctionTagsResponse?.tags?.some((tag) => tag.key === 'env')
      if (!hasEnvTag && typeof this.environment === 'string') {
        stepFunctionTagsToAdd.push({
          key: 'env',
          value: this.environment,
        })
      } else if (!hasEnvTag && this.environment === undefined) {
        this.context.stdout.write('\n[Error] --env is required when a Step Function has no env tag\n')

        return 1
      }

      if (
        !listStepFunctionTagsResponse?.tags?.some((tag) => tag.key === 'service') &&
        typeof this.service === 'string'
      ) {
        stepFunctionTagsToAdd.push({key: 'service', value: this.service})
      }

      // set version tag if it changed
      if (
        !listStepFunctionTagsResponse?.tags?.some(
          (tag) => tag.key === TAG_VERSION_NAME && tag.value === `v${cliVersion}`
        )
      ) {
        stepFunctionTagsToAdd.push({key: TAG_VERSION_NAME, value: `v${cliVersion}`})
      }

      if (stepFunctionTagsToAdd.length > 0) {
        await tagResource(stepFunctionsClient, stepFunctionArn, stepFunctionTagsToAdd, this.context, this.dryRun)
        hasChanges = true
      }

      const stateMachineName = describeStateMachineCommandOutput.name!
      const subscriptionFilterName = buildSubscriptionFilterName(stateMachineName)

      const logLevel = describeStateMachineCommandOutput.loggingConfiguration?.level

      if (logLevel === 'OFF') {
        // if step function logging is disabled, create a log group, subscribe the forwarder to it, and enable step function logging to the created log group
        const logGroupName = buildLogGroupName(stateMachineName, this.environment)
        await createLogGroup(cloudWatchLogsClient, logGroupName, stepFunctionArn, this.context, this.dryRun)

        await putSubscriptionFilter(
          cloudWatchLogsClient,
          this.forwarderArn,
          subscriptionFilterName,
          logGroupName,
          stepFunctionArn,
          this.context,
          this.dryRun
        )

        const logGroupArn = buildArn(
          arnObject.partition,
          'logs',
          arnObject.region,
          arnObject.accountId,
          'log-group',
          `${logGroupName}:*`
        )

        // Create Logs Access policy
        await createLogsAccessPolicy(
          iamClient,
          describeStateMachineCommandOutput,
          stepFunctionArn,
          this.context,
          this.dryRun
        )

        // Attach policy to state machine IAM role
        await attachPolicyToStateMachineIamRole(
          iamClient,
          describeStateMachineCommandOutput,
          arnObject.accountId,
          stepFunctionArn,
          this.context,
          this.dryRun
        )

        // IAM policy on step function role should include log permissions now
        await enableStepFunctionLogs(
          stepFunctionsClient,
          describeStateMachineCommandOutput,
          logGroupArn,
          stepFunctionArn,
          this.context,
          this.dryRun
        )
        hasChanges = true
      } else {
        // if step function logging is enabled, subscribe the forwarder to the log group in the step function logging config
        const logGroupArn = getStepFunctionLogGroupArn(describeStateMachineCommandOutput)
        if (logGroupArn === undefined) {
          this.context.stdout.write('\n[Error] Unable to get Log Group arn from Step Function logging configuration\n')

          return 1
        }
        const logGroupName = parseArn(logGroupArn).resourceName

        // update step function logging config to have logLevel `ALL` and includeExecutionData `true` if not already configured
        const includeExecutionData = describeStateMachineCommandOutput.loggingConfiguration?.includeExecutionData
        if (logLevel !== 'ALL' || !includeExecutionData) {
          await enableStepFunctionLogs(
            stepFunctionsClient,
            describeStateMachineCommandOutput,
            logGroupArn,
            stepFunctionArn,
            this.context,
            this.dryRun
          )
          hasChanges = true
        }
        await putSubscriptionFilter(
          cloudWatchLogsClient,
          this.forwarderArn,
          subscriptionFilterName,
          logGroupName,
          stepFunctionArn,
          this.context,
          this.dryRun
        )
        hasChanges = true
      }
    }
    if (!hasChanges) {
      this.context.stdout.write(`\nNo change is applied.\n `)
    }

    return 0
  }
}

InstrumentStepFunctionsCommand.addPath('stepfunctions', 'instrument')

InstrumentStepFunctionsCommand.addOption('dryRun', Command.Boolean('-d,--dry-run'))
InstrumentStepFunctionsCommand.addOption('environment', Command.String('-e,--env'))
InstrumentStepFunctionsCommand.addOption('forwarderArn', Command.String('--forwarder'))
InstrumentStepFunctionsCommand.addOption('service', Command.String('--service'))
InstrumentStepFunctionsCommand.addOption('stepFunctionArns', Command.Array('-s,--step-function'))
