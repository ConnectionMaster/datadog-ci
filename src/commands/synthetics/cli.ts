import {RunTestCommand} from './command'

module.exports = [RunTestCommand]
