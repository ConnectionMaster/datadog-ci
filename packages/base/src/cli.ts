/* eslint-disable quote-props */
import {CommandClass} from 'clipanion'

import {UploadCommand} from './commands/git-metadata/upload'
import {CheckCommand} from './commands/plugin/check-command'
import {DeployTestsCommand} from './commands/synthetics/deploy-tests-command'
import {ImportTestsCommand} from './commands/synthetics/import-tests-command'
import {RunTestsCommand} from './commands/synthetics/run-tests-command'
import {UploadApplicationCommand} from './commands/synthetics/upload-application-command'

// prettier-ignore
export const commands: Record<string, CommandClass[]> = {
  'git-metadata': [UploadCommand],
  'plugin': [CheckCommand],
  'synthetics': [RunTestsCommand, DeployTestsCommand, UploadApplicationCommand, ImportTestsCommand],
}
