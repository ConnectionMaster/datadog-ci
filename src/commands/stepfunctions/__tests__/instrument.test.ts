import {Cli} from 'clipanion/lib/advanced'

import {InstrumentStepFunctionsCommand} from '../instrument'

import {describeStateMachineFixture, stepFunctionTagListFixture} from './fixtures/aws-resources'
import {contextFixture, testContext} from './fixtures/cli'

jest.mock('../../../../package.json', () => ({version: '2.0.0'}))

describe('stepfunctions instrument test', () => {
  let cli: Cli
  let aws: any
  let helpers: any

  beforeAll(() => {
    helpers = require('../helpers')
    helpers.applyChanges = jest.fn().mockImplementation(() => false)

    cli = new Cli()
    cli.register(InstrumentStepFunctionsCommand)
  })

  let context: testContext
  beforeEach(() => {
    aws = require('../awsCommands')
    context = contextFixture()

    // different function responses may be needed depending on the test
    const describeStateMachineCommandOutput = describeStateMachineFixture()
    aws.describeStateMachine = jest.fn().mockImplementation(() => describeStateMachineCommandOutput)

    const stepFunctionTagList = [{key: 'env', value: 'test'}]
    aws.listTagsForResource = jest.fn().mockImplementation(() => ({tags: stepFunctionTagList}))

    aws.tagResource = jest.fn().mockImplementation(() => ({}))
    aws.putSubscriptionFilter = jest.fn().mockImplementation(() => ({}))
    aws.createLogsAccessPolicy = jest.fn().mockImplementation(() => ({}))
    aws.attachPolicyToStateMachineIamRole = jest.fn().mockImplementation(() => ({}))
    aws.enableStepFunctionLogs = jest.fn().mockImplementation(() => ({}))
    aws.putSubscriptionFilter = jest.fn().mockImplementation(() => ({}))
  })

  describe('paramater validation', () => {
    test('errors if forwarder arn is not set', async () => {
      const exitCode = await cli.run(['stepfunctions', 'instrument'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] --forwarder is required')
    })

    test('errors if forwarder arn is invalid', async () => {
      const exitCode = await cli.run(['stepfunctions', 'instrument', '--forwarder', 'arn:'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] invalid arn format for --forwarder arn:')
    })

    test('removes duplicate step function arns', async () => {
      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(aws.describeStateMachine).toHaveBeenCalledTimes(1)
    })

    test('errors if no step function arn', async () => {
      const exitCode = await cli.run(['stepfunctions', 'instrument'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] must specify at least one --step-function')
    })

    test('errors if any step function arn is invalid', async () => {
      const exitCode = await cli.run(['stepfunctions', 'instrument', '--step-function', 'arn:'], context)

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] invalid arn format for --step-function')
    })

    test('errors if no env tag on step function and env parameter not set', async () => {
      aws.listTagsForResource = jest.fn().mockImplementation(() => ({tags: []}))

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch('[Error] --env is required when a Step Function has no env tag')
    })
  })

  describe('step function tags', () => {
    test('sets dd_sls_ci tag if not already set', async () => {
      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--env',
          'test',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })

    test('sets dd_sls_ci tag if changed', async () => {
      const stepFunctionTagList = stepFunctionTagListFixture([{key: 'dd_sls_ci', value: 'v1.0.0'}])
      aws.listTagsForResource = jest.fn().mockImplementation(() => ({tags: stepFunctionTagList}))

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--env',
          'test',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })

    test('sets env tag if not already set', async () => {
      aws.listTagsForResource = jest.fn().mockImplementation(() => ({tags: [{key: 'dd_sls_ci', value: 'v2.0.0'}]}))

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--env',
          'test',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })

    test('sets service tag if not already set', async () => {
      const stepFunctionTagList = stepFunctionTagListFixture([{key: 'dd_sls_ci', value: 'v2.0.0'}])
      aws.listTagsForResource = jest.fn().mockImplementation(() => ({tags: stepFunctionTagList}))

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
          '--env',
          'test',
          '--service',
          'test-service',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })
  })

  describe('step function logging disabled', () => {
    test('creates step function log group, subscribes it to forwarder, and adds it to step function logging config', async () => {
      const loggingConfiguration = {
        level: 'OFF',
        includeExecutionData: false,
      }
      const stepFunction = describeStateMachineFixture({loggingConfiguration})
      aws.describeStateMachine = jest.fn().mockImplementation(() => stepFunction)

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })
  })

  describe('step function logging enabled', () => {
    test('subscribes log group in step function logging config to forwarder', async () => {
      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(0)
      expect(context.toString()).toMatchSnapshot()
    })
  })

  describe('aws error handling', () => {
    test('errors if unable to fetch step function', async () => {
      aws.describeStateMachine = jest.fn().mockImplementation(() => {
        throw new Error('mock describeStateMachine error')
      })

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch(
        '[Error] mock describeStateMachine error. Unable to describe state machine arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction'
      )
    })

    test('errors if unable to fetch step function tags', async () => {
      aws.listTagsForResource = jest.fn().mockImplementation(() => {
        throw new Error('mock listTagsForResource error')
      })

      const exitCode = await cli.run(
        [
          'stepfunctions',
          'instrument',
          '--forwarder',
          'arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder',
          '--step-function',
          'arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction',
        ],
        context
      )

      expect(exitCode).toBe(1)
      expect(context.toString()).toMatch(
        '[Error] mock listTagsForResource error. Unable to fetch tags for Step Function arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction'
      )
    })
  })
})
