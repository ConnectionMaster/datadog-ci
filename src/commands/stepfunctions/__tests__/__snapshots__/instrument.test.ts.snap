// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step-functions instrument step function logging disabled creates step function log group, subscribes it to forwarder, and adds it to step function logging config 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"dd_sls_ci\\",
      \\"value\\": \\"v2.0.0\\"
    }
  ]
}
createLogGroup ->
{
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs\\"
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs\\"
}
updateStateMachine ->
{
  \\"loggingConfiguration\\": {
    \\"level\\": \\"OFF\\",
    \\"includeExecutionData\\": false
  }
}
--->
{
  \\"loggingConfiguration\\": {
    \\"destinations\\": [
      {
        \\"cloudWatchLogsLogGroup\\": {
          \\"logGroupArn\\": \\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs:*\\"
        }
      }
    ],
    \\"level\\": \\"ALL\\",
    \\"includeExecutionData\\": true
  }
}
"
`;

exports[`step-functions instrument step function logging enabled subscribes log group in step function logging config to forwarder 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"dd_sls_ci\\",
      \\"value\\": \\"v2.0.0\\"
    }
  ]
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
"
`;

exports[`step-functions instrument step function tags sets dd_sls_ci tag if changed 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"dd_sls_ci\\",
      \\"value\\": \\"v2.0.0\\"
    }
  ]
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
"
`;

exports[`step-functions instrument step function tags sets dd_sls_ci tag if not already set 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"dd_sls_ci\\",
      \\"value\\": \\"v2.0.0\\"
    }
  ]
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
"
`;

exports[`step-functions instrument step function tags sets env tag if not already set 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"env\\",
      \\"value\\": \\"test\\"
    }
  ]
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
"
`;

exports[`step-functions instrument step function tags sets service tag if not already set 1`] = `
"
Will apply the following changes:

Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction
tagResource ->
{
  \\"resourceArn\\": \\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",
  \\"tags\\": [
    {
      \\"key\\": \\"service\\",
      \\"value\\": \\"test-service\\"
    }
  ]
}
putSubscriptionFilter ->
{
  \\"destinationArn\\": \\"arn:aws:lambda:us-east-1:000000000000:function:DatadogForwarder\\",
  \\"filterName\\": \\"ExampleStepFunctionLogGroupSubscription\\",
  \\"filterPattern\\": \\"\\",
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs-test\\"
}
"
`;

exports[`stepfunctions instrument test step function logging disabled creates step function log group, subscribes it to forwarder, and adds it to step function logging config 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"OFF\\",\\"includeExecutionData\\":false},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"env\\",\\"value\\":\\"test\\"}]}

Will apply the following change:
==============================
Changes for arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction

CreateLogGroup: 
{
  \\"logGroupName\\": \\"/aws/vendedlogs/states/ExampleStepFunction-Logs\\"
}

4444444444444444444444
"
`;

exports[`stepfunctions instrument test step function logging enabled subscribes log group in step function logging config to forwarder 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"ALL\\",\\"includeExecutionData\\":true,\\"destinations\\":[{\\"cloudWatchLogsLogGroup\\":{\\"logGroupArn\\":\\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs-test:*\\"}}]},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"env\\",\\"value\\":\\"test\\"}]}

No change is applied.
 "
`;

exports[`stepfunctions instrument test step function tags sets dd_sls_ci tag if changed 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"ALL\\",\\"includeExecutionData\\":true,\\"destinations\\":[{\\"cloudWatchLogsLogGroup\\":{\\"logGroupArn\\":\\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs-test:*\\"}}]},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"env\\",\\"value\\":\\"test\\"},{\\"key\\":\\"dd_sls_ci\\",\\"value\\":\\"v1.0.0\\"}]}

No change is applied.
 "
`;

exports[`stepfunctions instrument test step function tags sets dd_sls_ci tag if not already set 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"ALL\\",\\"includeExecutionData\\":true,\\"destinations\\":[{\\"cloudWatchLogsLogGroup\\":{\\"logGroupArn\\":\\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs-test:*\\"}}]},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"env\\",\\"value\\":\\"test\\"}]}

No change is applied.
 "
`;

exports[`stepfunctions instrument test step function tags sets env tag if not already set 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"ALL\\",\\"includeExecutionData\\":true,\\"destinations\\":[{\\"cloudWatchLogsLogGroup\\":{\\"logGroupArn\\":\\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs-test:*\\"}}]},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"dd_sls_ci\\",\\"value\\":\\"v2.0.0\\"}]}

No change is applied.
 "
`;

exports[`stepfunctions instrument test step function tags sets service tag if not already set 1`] = `
"
{\\"$metadata\\":{},\\"stateMachineArn\\":\\"arn:aws:states:us-east-1:000000000000:stateMachine:ExampleStepFunction\\",\\"name\\":\\"ExampleStepFunction\\",\\"definition\\":\\"\\",\\"loggingConfiguration\\":{\\"level\\":\\"ALL\\",\\"includeExecutionData\\":true,\\"destinations\\":[{\\"cloudWatchLogsLogGroup\\":{\\"logGroupArn\\":\\"arn:aws:logs:us-east-1:000000000000:log-group:/aws/vendedlogs/states/ExampleStepFunction-Logs-test:*\\"}}]},\\"roleArn\\":\\"arn:aws:iam::000000000000:role/ExampleStepFunctionRole\\",\\"type\\":\\"STANDARD\\",\\"creationDate\\":\\"2023-03-08T00:00:00.000Z\\"}

{\\"tags\\":[{\\"key\\":\\"env\\",\\"value\\":\\"test\\"},{\\"key\\":\\"dd_sls_ci\\",\\"value\\":\\"v2.0.0\\"}]}

No change is applied.
 "
`;
